src/data/multiset/fintype.lean

```diff
@@ -0,0 +1,203 @@
+begin
+  split,
+  { rintro ⟨⟨x, i, hx⟩, h⟩,
+    exact ⟨x, i, hx, h⟩, },
+  { rintro ⟨x, i, hx, h⟩,
+    exact ⟨⟨x, i, hx⟩, h⟩, },
+end
+
+instance {α : Type*} [decidable_eq α] (m : multiset α) : fintype {p : α × ℕ | p.2 < m.count p.1} :=
+fintype.of_finset
+(m.to_finset.bUnion (λ x, (finset.range (m.count x)).map ⟨prod.mk x, prod.mk.inj_left x⟩))
+begin
+  rintro ⟨x, i⟩,
+  simp only [finset.mem_bUnion, multiset.mem_to_finset, finset.mem_map, finset.mem_range,
+    function.embedding.coe_fn_mk, prod.mk.inj_iff, exists_prop, exists_eq_right_right,
+    set.mem_set_of_eq, and_iff_right_iff_imp],
+  exact λ h, multiset.count_pos.mp (pos_of_gt h),
+end
+
+/-- The embedding from a multiset into `α × ℕ` where the second coordinate enumerates repeats. -/
+@[simps]
+def multiset.coe_embedding {α : Type*} [decidable_eq α] (m : multiset α) :
+  m ↪ α × ℕ :=
+{ to_fun := λ x, (x, x.2.1),
+  inj' := begin
+    rintro ⟨x, i, hi⟩ ⟨y, j, hj⟩,
+    simp only [prod.mk.inj_iff, sigma.mk.inj_iff, and_imp, multiset.coe_eq],
+    rintro rfl rfl,
+    exact ⟨rfl, heq.rfl⟩
+  end }
+
+/-- Construct a finset whose elements enumerate the elements of the multiset `m`.
+The `ℕ` component is used to differentiate between equal elements: if `x` appears `n` times
+then `(x, 0)`, ..., `(x, n-1)` appear in the `finset`. -/
+def multiset.to_enum_finset {α : Type*} [decidable_eq α] (m : multiset α) : finset (α × ℕ) :=
+{p : α × ℕ | p.2 < m.count p.1}.to_finset
+
+@[simp] lemma multiset.mem_to_enum_finset {α : Type*} [decidable_eq α]
+  (m : multiset α) (p : α × ℕ) : p ∈ m.to_enum_finset ↔ p.2 < m.count p.1 :=
+set.mem_to_finset
+
+@[mono]
+lemma multiset.to_enum_finset_mono {α : Type*} [decidable_eq α] {m₁ m₂ : multiset α}
+  (h : m₁ ≤ m₂) : m₁.to_enum_finset ⊆ m₂.to_enum_finset :=
+begin
+  intro,
+  simp only [multiset.mem_to_enum_finset],
+  exact gt_of_ge_of_gt (multiset.le_iff_count.mp h a.1),
+end
+
+@[simp] lemma multiset.to_enum_finset_subset_iff {α : Type*} [decidable_eq α] {m₁ m₂ : multiset α} :
+  m₁.to_enum_finset ⊆ m₂.to_enum_finset ↔ m₁ ≤ m₂ :=
+begin
+  refine ⟨λ h, _, multiset.to_enum_finset_mono⟩,
+  rw multiset.le_iff_count,
+  intro x,
+  by_cases hx : x ∈ m₁,
+  { apply nat.le_of_pred_lt,
+    have : (x, m₁.count x - 1) ∈ m₁.to_enum_finset,
+    { rw multiset.mem_to_enum_finset,
+      exact nat.pred_lt (ne_of_gt (multiset.count_pos.mpr hx)), },
+    simpa only [multiset.mem_to_enum_finset] using h this, },
+  { simp [hx] },
+end
+
+/-- Another way to coerce a `multiset` to a type is to go through `m.to_enum_finset` and coerce
+that `finset` to a type. -/
+@[simps]
+def multiset.coe_equiv {α : Type*} [decidable_eq α] (m : multiset α) :
+  m ≃ m.to_enum_finset :=
+{ to_fun := λ x, ⟨m.coe_embedding x, by { rw multiset.mem_to_enum_finset, exact x.2.2 }⟩,
+  inv_fun := λ x, ⟨x.1.1, x.1.2, by { rw ← multiset.mem_to_enum_finset, exact x.2 }⟩,
+  left_inv := by { rintro ⟨x, i, h⟩, refl },
+  right_inv := by {rintro ⟨⟨x, i⟩, h⟩, refl } }
```

Comment:
Should probably have a `simp` lemma that `m.coe_equiv.to_embedding = m.coe_embedding`.